docs/permission/permission-framework-for-plugin-authors.md

```diff
@@ -0,0 +1,508 @@
+
+Create `plugins/todo-list-backend/src/service/rules.ts`
+
+```diff
+  import { makeCreatePermissionRule } from '@backstage/plugin-permission-node';
+- import { Todo } from './todos';
++ import { Todo, TodoFilter } from './todos';
+
+  const createTodoListPermissionRule = makeCreatePermissionRule<
+    Todo,
+-   undefined
++   TodoFilter
+  >();
+
+  export const isOwner = createTodoListPermissionRule({
+    name: 'IS_OWNER',
+    description: 'Should allow only if the todo belongs to the user',
+    apply: (resource, userId) => {
+      return resource.author === userId;
+    },
+    toQuery: userId => {
+-     throw new Error('toQuery not implemented');
++     return resource => resource.author === userId;
+    },
+  });
+```
+
+`plugins/todo-list-backend/src/service/router.ts`
+
+```diff
+
+  import {
+    todosListCreate,
+    todosListUpdate,
++   todosListRead,
+    TODOS_LIST_RESOURCE_TYPE,
+  } from './permissions';
+
+  router.get('/todos', async (req, res) => {
+    const token = IdentityClient.getBearerToken(req.header('authorization'));
++   const decision = (
++     await permissions.authorize([{ permission: todosListRead }], {
++       token,
++     })
++   )[0];
++   if (decision.result === AuthorizeResult.DENY) {
++     throw new NotAllowedError('Unauthorized');
++   }
+
++   if (decision.result === AuthorizeResult.CONDITIONAL) {
++     const conditionTransformer: ConditionTransformer<TodoFilter> =
++       createConditionTransformer(rules);
++     const filter = conditionTransformer(decision.conditions);
++     res.json(getAll(filter));
++     return;
++   }
+
+    res.json(getAll());
+  });
+```
+
+In this case, we are not passing any `resourceRef` when invoking `permissions.authorize()`.
+
+Since there is no `resourceRef` and the permission policy is returning a conditional response, the permission framework can't take a decision
+on its own and it's expecting you to take care of this case.
+This makes sense, since we would need to authorize each one of the todo items.
+
+We could implement this in a more efficient way.
+
+In case all the items are stored in a database. We could transform each permission result in the proper database query, letting the database do the filtering.
+
+In case it's ok for you to proceed in a more simple approach, it's still possible to use the result-by-result authorization approach mentioned at the beginning of the section.
+
+Fortunately, our todo service is smart enough and lets us pass an optional `filter` function when invoking `getAll()`.
+This is exactly what the `toQuery` present when defining each rule does.
+
+In this particular example, the `isOwner` rule returns a function (expected by `getAll` method). But there is no constraint regarding the shape of the returned object: any type of data can be returned.
+
+If you want to know more about how existing plugins integrate with the permission framework, check `catalog-backend` here.
```

Comment:
Because of the way in which we implemented the catalog backend integration, it may be a bit too heavyweight to point plugin authors as an example - this was one of the things discussed when designing the integration implementation, I think: that the way in which we tackled catalog may not be representative of what most plugin authors need.